sympy/core/expr.py

```diff
@@ -861,11 +863,21 @@ def _eval_interval(self, x, a, b):
                 domain = Interval(a, b)
             else:
                 domain = Interval(b, a)
-            singularities = list(solveset(self.cancel().as_numer_denom()[1], x, domain = domain))
+            # check the singularities of self within the interval
+            singularities = solveset(self.cancel().as_numer_denom()[1], x,
+                domain=domain)
+            for logterm in self.atoms(log):
+                singularities = singularities | solveset(logterm.args[0], x,
```

Comment:
@jksuom @normalhuman I am not sure though but solving for `logterm.args[0]` will only give specific points but we need all those points that make the term inside `log` negative like `log(x - 3)` here will give `{3}` as the branch point but I guess all the values of `x <= 3` are not applicable
I have tried to implement something like [here](https://github.com/sympy/sympy/pull/14792/files#diff-eec0422923e8f100745c015cd8fdd6cfR985) to remove singularities for solving logarithmic eqautions, have a look.

Reply:
The functions to be evaluated are not assumed to be real-valued. Hence logarithms having negative arguments should also be allowed.

Reply:
Thanks for the clarification.

Reply:
Thanks for the clarification.
What if we define `x` to be real, does this holds?
Actually, I was hoping to create a generalized function to determine logarithmic singularities (seeing its usage here as well as in solving the logarithmic equations (in `_transolve`)). What do you suggest?

Reply:
- > .. a singularity with slow growth like O(|log(x)|)

I didn't understand this correctly.

- One thing we can do to get the singularity of real functions (log) is by returning `Interval(-oo, b)`, where `b` is the branch point For complex ones just {b}. Will this be a good idea?

- Also why does `calculus.singularities` not solve by specifying the domain based on the assumption of the symbol? I may be wrong on this but I doubt the following:

```
>>> x = symbols('x', real=True)
>>> singularities(1/(x**2 + 1), x)
{-I, I}      # why the result is imaginary (if we want x to have real values), it should be `EmptySet()`

>>> x = symbols('x', real=False)
>>> singularities(1/(x**2 + 1), x)
{-I, I}
```

Reply:
> why the result is imaginary (if we want x to have real values)

If you look a the code, you see that `singularities` is based on `solveset` that apparently does not care about assumptions like `real=True` (it uses `domain` instead).

From the examples, it looks like the assumptions were planned to have a role but that was not implemented.

Reply:
> does not care about assumptions like real=True (it uses domain instead).

But we need to pass the domain when calling `solveset`. As of now, it is `Complex` by default.

Reply:
It looks like `domain` should be passed to `solveset`.

Reply:
Should I make PR regarding this?

Reply:
That might be useful. You could look at `symbol.is_real`.